Kernel.cc

```diff
@@ -1367,3 +1367,144 @@ short Kernel::findIndexNode( char * path , IndexNode& inode )
 }
 
+
+int Kernel::link(char *oldpath, char *newpath) {
+    	char dirname[1024];
+	memset(dirname, '\0', 1024);
+	strcpy(dirname, "/" );
+	int flags = O_WRONLY ; // ???
+	FileDescriptor * fileDescriptor = NULL;
+	FileSystem * fileSystem = openFileSystems;
+
+      	// get the full path
+	char *fullPath = getFullPath(oldpath);
+	IndexNode sourceIndexNode;
+	short sourceInodeNumber = findIndexNode(fullPath, sourceIndexNode);
+	cout << "size: " << sourceIndexNode.getSize() << endl;
+	
+	char *fullNewPath = getFullPath(newpath);
+	IndexNode newNode;
+	short newInodeNumber = findIndexNode(fullNewPath, newNode);
+	
+	fullNewPath = getFullPath(newpath);
+	char * token = NULL;
+	cout << fullNewPath << endl;
+	token = strtok(fullNewPath, "/");
+	char name[512];// = "." ; // start at root node
+	memset(name, '\0', 512);
+
+	cout << token << endl;
+
+	while(1) {
+	    if(token != NULL) {
+		memset(name, '\0', 512);
+		strcpy(name, token);
+	    } else
+		break;
+	    token = strtok(NULL, "/");
+	    if(token != NULL) {
+		strcat(dirname, name);	
+		strcat(dirname, "/");
+	    }
+	}
+	cout << "dirrrrr: " << dirname << endl;
+	
+	if(newInodeNumber < 0) {
+	    // Copy the source inode over
+	    fileSystem->readIndexNode(&sourceIndexNode, sourceInodeNumber);
+	    cout << "size: " << sourceIndexNode.getSize() << endl;
+
+	    // Increment nlinks
+	    int nlinks = sourceIndexNode.getNlink();
+	    nlinks++;
+	    sourceIndexNode.setNlink(nlinks);
+
+	    // Write inode
+	    fileSystem->writeIndexNode(&sourceIndexNode, sourceInodeNumber);
+	    
+	    fileDescriptor = new FileDescriptor(fileSystem, sourceIndexNode , flags);
+	    // assign inode for the new file
+	    fileDescriptor->setIndexNodeNumber(sourceInodeNumber);
+
+	    // open the directory
+	    // instead of a name for the dir
+	    cout << "dirname: " << dirname << endl;
+	    int dir = open(dirname , O_RDWR);
+	    if( dir < 0 ) {
+		perror(PROGRAM_NAME);
+		cout << PROGRAM_NAME << ": unable to open directory for writing" << endl;
+		return -1;
+	    }
+
+	    // scan past the directory entries less than the current entry
+	    // and insert the new element immediately following
+	    int status = 0;
+	    DirectoryEntry newDirectoryEntry(sourceInodeNumber, name);
+	    cout << "Name: " << name << " Inode: " << sourceInodeNumber << endl;
+	    DirectoryEntry currentDirectoryEntry;
+	    while(true)
+	    {
+		// read an entry from the directory
+		status = readdir(dir, currentDirectoryEntry);
+		cout << "Dir is: " << dir << " status: " << status << endl;
+		if(status < 0) {
+		    cout << PROGRAM_NAME << ": error reading directory in creat";
+		    exit( EXIT_FAILURE ) ;
+		}
+		else if( status == 0 ) {
+		    // if no entry read, write the new item at the current 
+		    // location and break
+		    writedir(dir , newDirectoryEntry);
+
+		    IndexNode note;
+		    char *funPath = getFullPath(newpath);
+		    short num = findIndexNode(funPath, note);
+		    cout << "num is: " << num << endl;
```

Comment:
This is the line that's needed for some reason. I  removed the test code I had hard coded before